File: browser/hns_vpn/android/hns_vpn_native_worker.cc
/* Copyright (c) 2021 The Hns Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "hns/browser/hns_vpn/android/hns_vpn_native_worker.h"

#include "base/android/jni_android.h"
#include "base/android/jni_array.h"
#include "base/android/jni_string.h"
#include "base/functional/bind.h"
#include "base/json/json_writer.h"
#include "base/values.h"
#include "hns/browser/hns_vpn/hns_vpn_service_factory.h"
#include "hns/build/android/jni_headers/HnsVpnNativeWorker_jni.h"
#include "hns/components/hns_vpn/browser/hns_vpn_service.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_manager.h"

using hns_vpn::HnsVpnService;

namespace {

HnsVpnService* GetHnsVpnService() {
  return hns_vpn::HnsVpnServiceFactory::GetForProfile(
      ProfileManager::GetActiveUserProfile()->GetOriginalProfile());
}

}  // namespace

namespace chrome {
namespace android {

HnsVpnNativeWorker::HnsVpnNativeWorker(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& obj)
    : weak_java_hns_vpn_native_worker_(env, obj), weak_factory_(this) {
  Java_HnsVpnNativeWorker_setNativePtr(env, obj,
                                         reinterpret_cast<intptr_t>(this));
}

HnsVpnNativeWorker::~HnsVpnNativeWorker() {}

void HnsVpnNativeWorker::Destroy(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  delete this;
}

void HnsVpnNativeWorker::GetAllServerRegions(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetAllServerRegions(
        base::BindOnce(&HnsVpnNativeWorker::OnGetAllServerRegions,
                       weak_factory_.GetWeakPtr()));
  }
}

void HnsVpnNativeWorker::OnGetAllServerRegions(
    const std::string& server_regions_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onGetAllServerRegions(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, server_regions_json),
      success);
}

void HnsVpnNativeWorker::GetTimezonesForRegions(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetTimezonesForRegions(
        base::BindOnce(&HnsVpnNativeWorker::OnGetTimezonesForRegions,
                       weak_factory_.GetWeakPtr()));
  }
}

void HnsVpnNativeWorker::OnGetTimezonesForRegions(
    const std::string& timezones_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onGetTimezonesForRegions(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, timezones_json), success);
}

void HnsVpnNativeWorker::GetHostnamesForRegion(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& region) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetHostnamesForRegion(
        base::BindOnce(&HnsVpnNativeWorker::OnGetHostnamesForRegion,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, region));
  }
}

void HnsVpnNativeWorker::OnGetHostnamesForRegion(
    const std::string& hostnames_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onGetHostnamesForRegion(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, hostnames_json), success);
}

void HnsVpnNativeWorker::GetWireguardProfileCredentials(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& subscriber_credential,
    const base::android::JavaParamRef<jstring>& public_key,
    const base::android::JavaParamRef<jstring>& hostname) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetWireguardProfileCredentials(
        base::BindOnce(&HnsVpnNativeWorker::OnGetWireguardProfileCredentials,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, subscriber_credential),
        base::android::ConvertJavaStringToUTF8(env, public_key),
        base::android::ConvertJavaStringToUTF8(env, hostname));
  }
}

void HnsVpnNativeWorker::OnGetWireguardProfileCredentials(
    const std::string& wireguard_profile_credentials_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onGetWireguardProfileCredentials(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(
          env, wireguard_profile_credentials_json),
      success);
}

void HnsVpnNativeWorker::VerifyCredentials(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& hostname,
    const base::android::JavaParamRef<jstring>& client_id,
    const base::android::JavaParamRef<jstring>& subscriber_credential,
    const base::android::JavaParamRef<jstring>& api_auth_token) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->VerifyCredentials(
        base::BindOnce(&HnsVpnNativeWorker::OnVerifyCredentials,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, hostname),
        base::android::ConvertJavaStringToUTF8(env, client_id),
        base::android::ConvertJavaStringToUTF8(env, subscriber_credential),
        base::android::ConvertJavaStringToUTF8(env, api_auth_token));
  }
}

void HnsVpnNativeWorker::OnVerifyCredentials(
    const std::string& verify_credentials_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onVerifyCredentials(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, verify_credentials_json),
      success);
}

void HnsVpnNativeWorker::InvalidateCredentials(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& hostname,
    const base::android::JavaParamRef<jstring>& client_id,
    const base::android::JavaParamRef<jstring>& subscriber_credential,
    const base::android::JavaParamRef<jstring>& api_auth_token) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->InvalidateCredentials(
        base::BindOnce(&HnsVpnNativeWorker::OnInvalidateCredentials,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, hostname),
        base::android::ConvertJavaStringToUTF8(env, client_id),
        base::android::ConvertJavaStringToUTF8(env, subscriber_credential),
        base::android::ConvertJavaStringToUTF8(env, api_auth_token));
  }
}

void HnsVpnNativeWorker::OnInvalidateCredentials(
    const std::string& invalidate_credentials_json,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onInvalidateCredentials(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, invalidate_credentials_json),
      success);
}

void HnsVpnNativeWorker::GetSubscriberCredential(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& product_type,
    const base::android::JavaParamRef<jstring>& product_id,
    const base::android::JavaParamRef<jstring>& validation_method,
    const base::android::JavaParamRef<jstring>& purchase_token,
    const base::android::JavaParamRef<jstring>& bundle_id) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetSubscriberCredential(
        base::BindOnce(&HnsVpnNativeWorker::OnGetSubscriberCredential,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, product_type),
        base::android::ConvertJavaStringToUTF8(env, product_id),
        base::android::ConvertJavaStringToUTF8(env, validation_method),
        base::android::ConvertJavaStringToUTF8(env, purchase_token),
        base::android::ConvertJavaStringToUTF8(env, bundle_id));
  }
}

void HnsVpnNativeWorker::GetSubscriberCredentialV12(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->GetSubscriberCredentialV12(
        base::BindOnce(&HnsVpnNativeWorker::OnGetSubscriberCredential,
                       weak_factory_.GetWeakPtr()));
  }
}

void HnsVpnNativeWorker::OnGetSubscriberCredential(
    const std::string& subscriber_credential,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onGetSubscriberCredential(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, subscriber_credential),
      success);
}

void HnsVpnNativeWorker::VerifyPurchaseToken(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& purchase_token,
    const base::android::JavaParamRef<jstring>& product_id,
    const base::android::JavaParamRef<jstring>& product_type,
    const base::android::JavaParamRef<jstring>& bundle_id) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->VerifyPurchaseToken(
        base::BindOnce(&HnsVpnNativeWorker::OnVerifyPurchaseToken,
                       weak_factory_.GetWeakPtr()),
        base::android::ConvertJavaStringToUTF8(env, purchase_token),
        base::android::ConvertJavaStringToUTF8(env, product_id),
        base::android::ConvertJavaStringToUTF8(env, product_type),
        base::android::ConvertJavaStringToUTF8(env, bundle_id));
  }
}

void HnsVpnNativeWorker::OnVerifyPurchaseToken(
    const std::string& json_response,
    bool success) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_HnsVpnNativeWorker_onVerifyPurchaseToken(
      env, weak_java_hns_vpn_native_worker_.get(env),
      base::android::ConvertUTF8ToJavaString(env, json_response), success);
}

jboolean HnsVpnNativeWorker::IsPurchasedUser(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    return hns_vpn_service->is_purchased_user();
  }

  return false;
}

void HnsVpnNativeWorker::ReloadPurchasedState(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->ReloadPurchasedState();
  }
}

void HnsVpnNativeWorker::ReportForegroundP3A(JNIEnv* env) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    // Reporting a new session to P3A functions.
    hns_vpn_service->RecordP3A(true);
  }
}

void HnsVpnNativeWorker::ReportBackgroundP3A(JNIEnv* env,
                                               jlong session_start_time_ms,
                                               jlong session_end_time_ms) {
  HnsVpnService* hns_vpn_service = GetHnsVpnService();
  if (hns_vpn_service) {
    hns_vpn_service->RecordAndroidBackgroundP3A(session_start_time_ms,
                                                  session_end_time_ms);
  }
}

static void JNI_HnsVpnNativeWorker_Init(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  new HnsVpnNativeWorker(env, jcaller);
}

}  // namespace android
}  // namespace chrome

File: android/java/org/chromium/chrome/browser/vpn/HnsVpnNativeWorker.java
/**
 * Copyright (c) 2021 The Hns Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.chromium.chrome.browser.vpn;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;

import java.util.ArrayList;
import java.util.List;

@JNINamespace("chrome::android")
public class HnsVpnNativeWorker {
    private long mNativeHnsVpnNativeWorker;
    private static final Object mLock = new Object();
    private static HnsVpnNativeWorker mInstance;

    private List<HnsVpnObserver> mObservers;

    public static HnsVpnNativeWorker getInstance() {
        synchronized (mLock) {
            if (mInstance == null) {
                mInstance = new HnsVpnNativeWorker();
                mInstance.init();
            }
        }
        return mInstance;
    }

    private HnsVpnNativeWorker() {
        mObservers = new ArrayList<HnsVpnObserver>();
    }

    private void init() {
        if (mNativeHnsVpnNativeWorker == 0) {
            HnsVpnNativeWorkerJni.get().init(this);
        }
    }

    @Override
    protected void finalize() {
        destroy();
    }

    private void destroy() {
        if (mNativeHnsVpnNativeWorker != 0) {
            HnsVpnNativeWorkerJni.get().destroy(mNativeHnsVpnNativeWorker, this);
            mNativeHnsVpnNativeWorker = 0;
        }
    }

    public void addObserver(HnsVpnObserver observer) {
        synchronized (mLock) {
            mObservers.add(observer);
        }
    }

    public void removeObserver(HnsVpnObserver observer) {
        synchronized (mLock) {
            mObservers.remove(observer);
        }
    }

    @CalledByNative
    private void setNativePtr(long nativePtr) {
        assert mNativeHnsVpnNativeWorker == 0;
        mNativeHnsVpnNativeWorker = nativePtr;
    }

    @CalledByNative
    public void onGetAllServerRegions(String jsonServerRegions, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onGetAllServerRegions(jsonServerRegions, isSuccess);
        }
    }

    @CalledByNative
    public void onGetTimezonesForRegions(String jsonTimezones, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onGetTimezonesForRegions(jsonTimezones, isSuccess);
        }
    }

    @CalledByNative
    public void onGetHostnamesForRegion(String jsonHostnames, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onGetHostnamesForRegion(jsonHostnames, isSuccess);
        }
    }

    @CalledByNative
    public void onGetWireguardProfileCredentials(
            String jsonWireguardProfileCredentials, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onGetWireguardProfileCredentials(jsonWireguardProfileCredentials, isSuccess);
        }
    }

    @CalledByNative
    public void onVerifyCredentials(String jsonVerifyCredentials, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onVerifyCredentials(jsonVerifyCredentials, isSuccess);
        }
    }

    @CalledByNative
    public void onInvalidateCredentials(String jsonInvalidateCredentials, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onInvalidateCredentials(jsonInvalidateCredentials, isSuccess);
        }
    }

    @CalledByNative
    public void onGetSubscriberCredential(String subscriberCredential, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onGetSubscriberCredential(subscriberCredential, isSuccess);
        }
    }

    @CalledByNative
    public void onVerifyPurchaseToken(String jsonResponse, boolean isSuccess) {
        for (HnsVpnObserver observer : mObservers) {
            observer.onVerifyPurchaseToken(jsonResponse, isSuccess);
        }
    }

    public void getAllServerRegions() {
        HnsVpnNativeWorkerJni.get().getAllServerRegions(mNativeHnsVpnNativeWorker);
    }

    public void getTimezonesForRegions() {
        HnsVpnNativeWorkerJni.get().getTimezonesForRegions(mNativeHnsVpnNativeWorker);
    }

    public void getHostnamesForRegion(String region) {
        HnsVpnNativeWorkerJni.get().getHostnamesForRegion(mNativeHnsVpnNativeWorker, region);
    }

    public void getWireguardProfileCredentials(
            String subscriberCredential, String publicKey, String hostname) {
        HnsVpnNativeWorkerJni.get().getWireguardProfileCredentials(
                mNativeHnsVpnNativeWorker, subscriberCredential, publicKey, hostname);
    }

    public void verifyCredentials(
            String hostname, String clientId, String subscriberCredential, String apiAuthToken) {
        HnsVpnNativeWorkerJni.get().verifyCredentials(mNativeHnsVpnNativeWorker, hostname,
                clientId, subscriberCredential, apiAuthToken);
    }

    public void invalidateCredentials(
            String hostname, String clientId, String subscriberCredential, String apiAuthToken) {
        HnsVpnNativeWorkerJni.get().invalidateCredentials(mNativeHnsVpnNativeWorker, hostname,
                clientId, subscriberCredential, apiAuthToken);
    }

    public void getSubscriberCredential(String productType, String productId,
            String validationMethod, String purchaseToken, String packageName) {
        HnsVpnNativeWorkerJni.get().getSubscriberCredential(mNativeHnsVpnNativeWorker,
                productType, productId, validationMethod, purchaseToken, packageName);
    }

    public void verifyPurchaseToken(
            String purchaseToken, String productId, String productType, String packageName) {
        HnsVpnNativeWorkerJni.get().verifyPurchaseToken(
                mNativeHnsVpnNativeWorker, purchaseToken, productId, productType, packageName);
    }

    // Desktop purchase methods
    public void reloadPurchasedState() {
        HnsVpnNativeWorkerJni.get().reloadPurchasedState(mNativeHnsVpnNativeWorker);
    }

    public boolean isPurchasedUser() {
        return HnsVpnNativeWorkerJni.get().isPurchasedUser(mNativeHnsVpnNativeWorker);
    }

    public void getSubscriberCredentialV12() {
        HnsVpnNativeWorkerJni.get().getSubscriberCredentialV12(mNativeHnsVpnNativeWorker);
    }

    public void reportBackgroundP3A(long sessionStartTimeMs, long sessionEndTimeMs) {
        HnsVpnNativeWorkerJni.get().reportBackgroundP3A(
                mNativeHnsVpnNativeWorker, sessionStartTimeMs, sessionEndTimeMs);
    }

    public void reportForegroundP3A() {
        HnsVpnNativeWorkerJni.get().reportForegroundP3A(mNativeHnsVpnNativeWorker);
    }

    @NativeMethods
    interface Natives {
        void init(HnsVpnNativeWorker caller);
        void destroy(long nativeHnsVpnNativeWorker, HnsVpnNativeWorker caller);
        void getAllServerRegions(long nativeHnsVpnNativeWorker);
        void getTimezonesForRegions(long nativeHnsVpnNativeWorker);
        void getHostnamesForRegion(long nativeHnsVpnNativeWorker, String region);
        void getWireguardProfileCredentials(long nativeHnsVpnNativeWorker,
                String subscriberCredential, String publicKey, String hostname);
        void verifyCredentials(long nativeHnsVpnNativeWorker, String hostname, String clientId,
                String subscriberCredential, String apiAuthToken);
        void invalidateCredentials(long nativeHnsVpnNativeWorker, String hostname,
                String clientId, String subscriberCredential, String apiAuthToken);
        void getSubscriberCredential(long nativeHnsVpnNativeWorker, String productType,
                String productId, String validationMethod, String purchaseToken,
                String packageName);
        void verifyPurchaseToken(long nativeHnsVpnNativeWorker, String purchaseToken,
                String productId, String productType, String packageName);
        void reloadPurchasedState(long nativeHnsVpnNativeWorker);
        boolean isPurchasedUser(long nativeHnsVpnNativeWorker);
        void getSubscriberCredentialV12(long nativeHnsVpnNativeWorker);
        void reportBackgroundP3A(
                long nativeHnsVpnNativeWorker, long sessionStartTimeMs, long sessionEndTimeMs);
        void reportForegroundP3A(long nativeHnsVpnNativeWorker);
    }
}
